modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1769,7 +1770,138 @@ def sort_by(
+                from each. It samples using the following algorithm:
+                    * If there are <= 100 rows in the dataframe, we pick quantiles over the entire dataframe.
+                    * If there are 100 < # of rows <= 1900, we pick quantiles over the first 100 rows, plus a sample
+                    of 5% of the remaining rows.
+                    * If there are > 1900 rows, we pick quantiles over a sample of 10% of all of the rows.
+                
+                These numbers are a heuristic. They were picked such that the size of the sample
+                in scenario 2 scales well. In other words, we picked A, k, and q such that:
+                    A + (len(df) - A)*k = q*len(df)
+                Where q is the proportion we sample in scenario 3, k is the proportion we sample
+                of remaining rows in scenario 2, and A is the threshold below which we just return
+                the entire dataframe, instead of sampling (scenario 1).
+                q = 0.1 and k = 0.05 were picked such that this formula holds for A = 100.
+                """
+                quantiles = [
+                    i / (NPartitions.get() * 2) for i in range(NPartitions.get() * 2)
+                ]
+                # Heuristic for a "small" df we will compute histogram over entirety of.
+                if len(df) <= 100:
+                    return np.quantile(df[columns[0]], quantiles)
+                # Heuristic for a "medium" df where we will include first 100 rows, and sample
+                # of remaining rows
+                if len(df) <= 1900:
+                    return np.quantile(
+                        np.concatenate(
+                            (
+                                df[columns[0]][:100].values,
+                                df[columns[0]][100:].sample(frac=0.05),
+                            )
+                        ),
+                        quantiles,
+                    )
+                return np.quantile(df[columns[0]].sample(frac=0.1), quantiles)
+
+            def pivot_func(samples):
+                """
+                Determine quantiles from the given samples.
+
+                This function takes as input the quantiles calculated over all partitions from 
+                `sample_func` defined above, and determines a final NPartitions.get() * 2 quantiles
+                to use to roughly sort the entire dataframe. It does so by collating all the samples
+                and computing NPartitions.get() * 2 quantiles for the overall set.
+                """
+                # We don't call `np.unique` on the samples, since if a quantile shows up in multiple
+                # partition's samples, this is probably an indicator of skew in the dataset, and we
+                # want our final partitions to take this into account.
+                all_pivots = np.array(samples).flatten()
```

Comment:
I previously had an `np.unique` here as a holdover from when I was directly passing this to the output (so the code looked like `return np.unique(np.sort(np.array(samples).flatten()))`, but have removed the `np.unique` since I figure if a quantile appears in multiple partitions' quantiles, thats probably a good indicator of skew in the overall dataset, and our final quantiles should take that into account.

I'm wondering if folks think this makes sense or if I should add the `np.unique` call back?

Reply:
@naren-ponder the tradeoff is a larger array to find quantiles over and the potential for duplicate quantiles, but I think its well worth it to properly model skew in the dataset. The other tradeoff is if the same quantiles appear in multiple datasets due to skew in the distribution of data across partitions, that over represents the cardinality of that data in the overall dataset (i.e. the skew doesn't actually exist in the overall dataset), but I think these are relatively minor pathological cases.